sdk/core/azure-core/CMakeLists.txt

```diff
@@ -59,6 +62,7 @@ create_code_coverage(core ${TARGET_NAME} "${TARGET_NAME}-test")
 
 target_include_directories(${TARGET_NAME} PUBLIC ${CURL_INCLUDE_DIRS})
 target_link_libraries(${TARGET_NAME} PRIVATE CURL::libcurl)
+target_link_libraries(${TARGET_NAME} INTERFACE nlohmann_json::nlohmann_json)
```

Comment:
Can you help me understand why we are picking `interface` here? 

I am reading this to see when to use when we would use that: https://cmake.org/cmake/help/v3.0/command/target_link_libraries.html

Is this the reason for using interface here?
> If your header files but not your source files include the headers of another library, then it is an INTERFACE dependency.

Reply:
I don't think it should be `INTERFACE`

Reply:
> We can leave out this keyword.

What are the semantics if a keyword is not provided?

Reply:
sure, INTERFACE means that the target/lib is not linked and become part of Azure Core. It is only set as to be linked together with anyone linking azure core.
So, when an end-user or a service links azure-core, CMake will translate to "link azure core and also link this json lib".
The difference with PUBLIC is that the library bytes are copied directly to azure-core. Then when azure core is linked, there is no extra libs linked (behind the scenes).

The benefit is that Azure Core will not grow in size by adding the entire Json lib (as with PUBLIC)

Usually services, or anyone creating a new library on top of azure core would use PRIVATE linking, meaning that the lib will only take the required bytes from the linked library without re-exposing its functionality. Meaning that if the new library is only using one method from JSON, only that method will be copied to the new library.

If a new library would like to expose the API from an inner library, (like if Storage would like to offer JSON api), they would link it as INTERFACE to avoid copying all inside core lib.

For the previous approach, where Storage team was including the header file directly as part of the sources, it was making Storage common to offer the entire JSON API to any customer (like doing PUBLIC linking).  I don't think we want to do that.

Reply:
btw, @Jinming-Hu , the link dependencies option keyword is mandatory. There's no default for it.

Reply:
> The difference with PUBLIC is that the library bytes are copied directly to azure-core

I seriously doubt that.

Reply:
> the link dependencies option keyword is mandatory. There's no default for it.

I don't think so. https://cmake.org/cmake/help/latest/command/target_link_libraries.html

Reply:
Really, I don't think it makes too much difference whether we use `INTERFACE` or `PUBLIC` here, or even leave it out.

Reply:
@Jinming-Hu , did you actually read that link?  
Make sure you select the 3.13 version (https://cmake.org/cmake/help/v3.13/command/target_link_libraries.html)
Then, notice that if we want to omit the KEYWORD (PUBLIC, INTERFACE, PRIVATE), the item we add must be either:
-  library target name
-  A full path to a library file
-  A plain library name
-  A link flag

But what we need is to link an `ALIAS library Target name`. There is even this note in there:
```
Items containing ::, such as Foo::Bar, are assumed to be IMPORTED or ALIAS library target names and will cause an error if no such target exists. See policy CMP0028.
```

Then there is a section describing how to link this by using the Keyword.

Take a look to the next links:
https://stackoverflow.com/questions/26037954/cmake-target-link-libraries-interface-dependencies
https://cmake.org/cmake/help/v3.0/manual/cmake-buildsystem.7.html#transitive-usage-requirements

There is more info about the intention of each KEYWORD. 

Think about it like "Azure CORE is an interface for the JSON lib for others to consume it through it. It is used as a bridge"